### PR TITLE
Requests tx in order.

### PR DESCRIPTION
**Pull request by karknu**
There is a problem with the way the TxSubmission protocol tracks which txs has been accepted. The only mechanics provided is a simple window size update. This means that if the server decides to fetch txs in an arbitrary order it is very hard for the client to keep track of which txid's it has sent to the client.

By requesting Tx in the order they where presented (if the server is interested in them) the client can easily keep track of a limited number of TXs (rather than all TXs sent since the start of the connection), and the server can verify that the client does indeed send newish txs.

Requesting tx in order also means that they get added to the mempool faster and thus propagate through the network quicker.

### CODE REVIEW

**Comment by karknu** [comment]
Fixes #2680 in a controversial way.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs [hunk 1]
```diff
@@ -253,11 +256,18 @@ txSubmissionInbound _tracer maxUnacked mpReader mpWriter _version =
...
+            txidsAndSizeSeq = Seq.fromList txids
+            txidsSet = Set.fromList (map fst txids)
+            unAckSet = Set.fromList $ toList $ unacknowledgedTxIds st
 
         unless (Seq.length txidsSeq <= fromIntegral reqNo) $
           throwM ProtocolErrorTxIdsNotRequested
 
+        -- We are requesting *new* txids from the peer. There shouldn't
+        -- be any overlap with txids that we still haven't processed.
+        unless (Set.intersection txidsSet unAckSet == Set.empty) $
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:268]: `Set.null`?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs [hunk 2]
```diff
@@ -350,12 +360,21 @@ txSubmissionInbound _tracer maxUnacked mpReader mpWriter _version =
...
         }
       where
+
+        -- TODO: Add to StrictSeq.
+        partition :: (a -> Bool) -> StrictSeq a -> (StrictSeq a, StrictSeq a)
+        partition cond = foldl' part (Seq.empty, Seq.empty)
+          where
+            part (ls, rs) e
+              | cond e    = ((Seq.|>) ls e, rs)
+              | otherwise = (ls, (Seq.|>) rs e)
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:370]: Alternatively:
```suggestion
        partition :: (a -> Bool) -> StrictSeq a -> (StrictSeq a, StrictSeq a)
        partition cond xs = (Seq.toStrict ls, Seq.toStrict rs)
          where
            (ls, rs) = LazySeq.partition cond xs
```
with: `import qualified Data.Sequence as LazySeq`
I'm not sure which one is more efficient.

**Comment by karknu** [comment]
This doesn't fix #2680 it just makes it harder to trigger.

**Pull request closed without merging**